feat(storage): add support for sessionStorage as identityStorage

nutgaard opened the pull request:
Proposed fix for #434

<!---
Thanks for contributing to the Amplitude TypeScript repository! 🎉

Please fill out the following sections to help us quickly review your pull request.
--->

### Summary

Adds support for sessionStorage as identityStorage

```ts
amplitude.init(<YOUR_API_KEY>, {
  identityStorage: 'sessionStorage',
});
```

### Checklist

* [x] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:  No

nutgaard commented:
Saw the comments from @kelvin-lu in #437; 
I've tried to address these [in a separate PR](https://github.com/nutgaard/Amplitude-TypeScript/pull/1) which atm is still just in my fork. 

**TLDR:** 
```typescript
import { Storage as AmplitudeStorage } from '@amplitude/analytics-types';

class BrowserStorage<T> implement AmplitudeStorage<T> {
  constructor(private storage?: Storage) {}
  
  // Rest of the class is similar to local-storage.ts and session-storage.ts
}

class LocalStorage<T> extends BrowserStorage<T> {
  constructor() {
    super(getGlobalScope()?.localStorage);
  }
}

class SessionStorage<T> extends BrowserStorage<T> {
  constructor() {
    super(getGlobalScope()?.sessionStorage);
  }
}
```

This is possible since `globalThis.localStorage` and `globalThis.sessionStorage` both implement a common interface `Storage` (not to be confused with the one specified in analytics-types). 


I've also added [console.error](https://github.com/nutgaard/Amplitude-TypeScript/pull/1/files#diff-92ee6a32e0df65f9af4cabaf9490f9176aa0dbc9d60f4931fb050d96377c1d02R36) where mentioned. But the error message itself could probably be improved.


If these changes looks like something you all would have included in this PR, then please let me know and I'll update the PR.

kevinpagtakhan commented:
> Saw the comments from @kelvin-lu in #437; I've tried to address these [in a separate PR](https://github.com/nutgaard/Amplitude-TypeScript/pull/1) which atm is still just in my fork.
> 
> **TLDR:**
> 
> ```ts
> import { Storage as AmplitudeStorage } from '@amplitude/analytics-types';
> 
> class BrowserStorage<T> implement AmplitudeStorage<T> {
>   constructor(private storage?: Storage) {}
>   
>   // Rest of the class is similar to local-storage.ts and session-storage.ts
> }
> 
> class LocalStorage<T> extends BrowserStorage<T> {
>   constructor() {
>     super(getGlobalScope()?.localStorage);
>   }
> }
> 
> class SessionStorage<T> extends BrowserStorage<T> {
>   constructor() {
>     super(getGlobalScope()?.sessionStorage);
>   }
> }
> ```
> 
> This is possible since `globalThis.localStorage` and `globalThis.sessionStorage` both implement a common interface `Storage` (not to be confused with the one specified in analytics-types).
> 
> I've also added [console.error](https://github.com/nutgaard/Amplitude-TypeScript/pull/1/files#diff-92ee6a32e0df65f9af4cabaf9490f9176aa0dbc9d60f4931fb050d96377c1d02R36) where mentioned. But the error message itself could probably be improved.
> 
> If these changes looks like something you all would have included in this PR, then please let me know and I'll update the PR.

@nutgaard yes please, let's put them together in the same PR. would like to get this in :)

nutgaard commented:
Saw the builds failed due to insufficient coverage, e.g the added `console.error` statement was not covered during test.

The following returns-statement was previously annotated with `/* istanbul ignore next */`. In order to avoid adding another ignore-annotation I've added a new test which covers this case (invalid json stored in storage) and removed the original annotation.
